crates/containerd-shim-wasm/src/sandbox/sync.rs

```diff
@@ -0,0 +1,117 @@
+use std::cell::OnceCell;
+use std::sync::{Arc, Condvar, Mutex};
+use std::time::Duration;
+
+/// A cell where we can wait (with timeout) for
+/// a value to be set
+pub struct WaitableCell<T> {
+    inner: Arc<WaitableCellImpl<T>>,
+}
+
+struct WaitableCellImpl<T> {
+    // Ideally we would just use a OnceLock, but it doesn't
+    // have the `wait` and `wait_timeout` methods, so we use
+    // a Condvar + Mutex pair instead.
+    // We can't guard the OnceCell **inside** the Mutex as
+    // that would produce ownership problems with returning
+    // `&T`. This is because the mutex doesn't know that we
+    // won't mutate the OnceCell once it's set.
+    mutex: Mutex<()>,
+    cvar: Condvar,
+    cell: OnceCell<T>,
+}
+
+// this is safe because access to cell guarded by the mutex
+unsafe impl<T> Send for WaitableCell<T> {}
+unsafe impl<T> Sync for WaitableCell<T> {}
+
+struct WaitableCellSetGuard<T, R: Into<T>, F: FnOnce() -> R> {
+    f: Option<F>,
+    cell: WaitableCell<T>,
+}
+
+impl<T> Default for WaitableCell<T> {
+    fn default() -> Self {
+        Self {
+            inner: Arc::new(WaitableCellImpl {
+                mutex: Mutex::new(()),
+                cvar: Condvar::new(),
+                cell: OnceCell::new(),
+            }),
+        }
+    }
+}
+
+impl<T> Clone for WaitableCell<T> {
+    fn clone(&self) -> Self {
+        let inner = self.inner.clone();
+        Self { inner }
+    }
+}
+
+impl<T> WaitableCell<T> {
+    /// Creates an empty WaitableCell.
+    pub fn new() -> Self {
+        Self::default()
+    }
+
+    /// Creates a WaitableCell with a value.
+    pub fn with_value(val: T) -> Self {
+        let cell = Self::default();
+        let _ = cell.set(val);
+        cell
+    }
+
+    /// Sets a value to the WaitableCell.
+    /// This method has no effect if the WaitableCell already has a value.
+    pub fn set(&self, val: impl Into<T>) -> Result<(), T> {
+        let val = val.into();
+        let _guard = self.inner.mutex.lock().unwrap();
+        let res = self.inner.cell.set(val);
+        self.inner.cvar.notify_all();
+        res
+    }
+
+    /// If the `WaitableCell` is empty when this guard is dropped, it will set with `val`.
+    pub fn set_guard(&self, val: impl Into<T>) -> impl Drop {
+        self.set_guard_with(move || val.into())
+    }
+
+    /// If the `WaitableCell` is empty when this guard is dropped, it will set with the result of `f`.
+    pub fn set_guard_with<R: Into<T>>(&self, f: impl FnOnce() -> R) -> impl Drop {
+        let cell = (*self).clone();
+        WaitableCellSetGuard { f: Some(f), cell }
+    }
+
+    /// Wait for the WaitableCell to be set a value.
+    pub fn wait(&self) -> &T {
+        let value = self.wait_timeout(None);
```

Comment:
When `wait_timeout` returns a `None` value, this could be dangerous to do `unwrap_unchecked()`.

Could you please add a `assert` statement to check the value is not `None` at runtime at least?